### PR TITLE
collectLegacyRoutes throws error in case invalid element is found

### PR DESCRIPTION
**Pull request by vinzscam**
## Hey, I just made a Pull Request!

`collectLegacyRoutes` throws error in case an invalid `<Route />` element is found or `<Route />` doesn't contain `path`.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/core-compat-api/src/collectLegacyRoutes.tsx [hunk 1]
```diff
@@ -186,7 +185,12 @@ export function collectLegacyRoutes(
...
         'core.mountPoint',
       );
-      if (!plugin || !path) {
+      if (path === undefined) {
+        throw new Error(
+          `<Route /> element with invalid path has been detected. Please make sure to pass the path's props`,
+        );
+      }
+      if (!plugin) {
         return;
```
Comment (vinzscam) [inline packages/core-compat-api/src/collectLegacyRoutes.tsx:194]: > [!NOTE]
I haven't thrown anything in case no plugin is detected to avoid getting troubles with react router's corner cases such as `<Navigate />`

File: packages/core-compat-api/src/collectLegacyRoutes.test.tsx [hunk 2]
```diff
@@ -279,4 +279,58 @@ describe('collectLegacyRoutes', () => {
...
+    const plugin = createPlugin({
+      id: 'test',
+    });
+    const routeRef = createRouteRef({ id: 'test' });
+    const Page = plugin.provide(
+      createRoutableExtension({
+        name: 'Test',
+        mountPoint: routeRef,
+        component: () =>
+          Promise.resolve(() => {
```
Comment (freben) [inline packages/core-compat-api/src/collectLegacyRoutes.test.tsx:293]: very matter of taste, feel free to ignore:

```suggestion
        component: async () => () => {
```

haha i know you're just being consistent with the others, but there's something about Promise.resolve that just bugs me :) same below

File: packages/core-compat-api/src/collectLegacyRoutes.tsx [hunk 3]
```diff
@@ -173,9 +173,8 @@ export function collectLegacyRoutes(
     (route: ReactNode) => {
       // TODO(freben): Handle feature flag and permissions framework wrapper elements
       if (!React.isValidElement(route) || route.type !== Route) {
-        return;
+        throw new Error('Invalid <Route /> element has been detected.');
```
Comment (freben) [inline packages/core-compat-api/src/collectLegacyRoutes.tsx:176]:
```suggestion
        throw new Error(`Invalid element inside FlatRoutes, expected Route but found ${route.type}.`);
```

at least in the code path where we know that it's a valid element? There's a little bit too little info in the message otherwise, to understand what went wrong and for us to help debug.

File: packages/core-compat-api/src/collectLegacyRoutes.tsx [hunk 4]
```diff
@@ -186,7 +185,12 @@ export function collectLegacyRoutes(
         routeElement,
         'core.mountPoint',
       );
-      if (!plugin || !path) {
+      if (path === undefined) {
+        throw new Error(
+          `<Route /> element with invalid path has been detected. Please make sure to pass the path's props`,
```
Comment (freben) [inline packages/core-compat-api/src/collectLegacyRoutes.tsx:190]:
```suggestion
          `Route element inside FlatRoutes had no path prop value given`,
```

or something like that?

**Review by freben — COMMENTED** [review]
Alright just suggesting some wording

File: packages/core-compat-api/src/collectLegacyRoutes.tsx (hunk 1 above)
Comment (Rugvip) [reply packages/core-compat-api/src/collectLegacyRoutes.tsx:194]: I think we should throw an error in this case too, and let users know that they can remove the route if needed. Something like `Route with path '...' has en element that can not be converted as it does not belong to a plugin. Make sure that the top-level React element of the element prop is an extension from a Backstage plugin, or remove the Route completely. See <link-to-app-migration-docs> for more info`. In the migration docs we can then mention that you should remove the `Navigation` element and replace it with configuration of the `app/routes` extension (which we should also add).

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice 😎 

Just a nit on the last error case

**Review by Rugvip — APPROVED** [review]
Nice! 👍 🎉 

Let's :shipit:
